sql: enable autovacuum_enabled storage param setting

otan opened the pull request:
This change involved a change to the syntax to be more in line with
PostgreSQL's gram.y for storage parameters, allowing most expression
values in the WITH clause. We further parse unresolved names as strings
to allow something like `autovacuum_enabled = off` to work.

Resolves #51818

Release note (sql change): Allow parsing of `CREATE TABLE ... WITH
(autovacuum_enabled = bool)`, which is a no-op operation.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52652)
<!-- Reviewable:end -->

rohany commented on pkg/sql/create_index.go:
```diff
@@ -34,14 +35,27 @@ type createIndexNode struct {
-	`pages_per_range`:                   storageParamUnimplemented,
-	`autosummarize`:                     storageParamUnimplemented,
+type indexStorageParamObserver struct {
+	indexDesc *descpb.IndexDescriptor
```
why does this need an indexDesc?

otan replied:
in a future state, you may alter the index descriptor based on the storage params.
(i will need to do this later, so...)

rohany replied:
I think that you should just add them once you need them

otan replied:
ok

rohany commented on pkg/sql/create_index.go:
```diff
@@ -243,6 +247,17 @@ func MakeIndexDescriptor(
 		return nil, err
 	}
+
+	paramObserver := indexStorageParamObserver{indexDesc: &indexDesc}
```
nit i'd just inline the creation of this observer

otan replied:
BUT MY HEAP
ok

rohany commented on pkg/sql/create_table.go:
```diff
@@ -73,47 +73,91 @@ type createTableRun struct {
-	storageParamUnimplemented
-)
+type tableStorageParamObserver struct {
+	tblDesc *sqlbase.MutableTableDescriptor
```
why does this need a table

otan replied:
in a future state, you may alter the table descriptor based on the storage params.
(e.g. you actually want to alter autovacuum_enabled to set some sort of bool on the descriptor to tell the opt not to collect stats)

otan committed:
sql: enable autovacuum_enabled storage param setting

This change involved a change to the syntax to be more in line with
PostgreSQL's gram.y for storage parameters, allowing most expression
values in the WITH clause. We further parse unresolved names as strings
to allow something like `autovacuum_enabled = off` to work.

Release note (sql change): Allow parsing of `CREATE TABLE ... WITH
(autovacuum_enabled = bool)`, which is a no-op operation.

otan commented:
bors r=rohany
